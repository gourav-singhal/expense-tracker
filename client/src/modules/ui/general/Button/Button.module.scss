.base {
  position: relative;
  display: inline-block;
  width: var(--inner-max-width);
  padding: var(--space-base);
  font-weight: 500;
  line-height: 130%;
  text-align: center;
  overflow-wrap: break-word;
  cursor: pointer;
  transition: background-color .3s ease-in-out, border-color .3s ease-in-out, color .3s ease-in-out;
  border: none;
  border-radius: var(--border-radius-base);

  &:active {
    transform: scale(.96);
  }
}

.wrapper {
  display: flex;
  align-items: center;
  justify-content: space-between;
}

.iconButton {
  width: max-content;
  padding: var(--space-xs);
}

.primary {
  background-color: var(--color-brand-yellow);

  &:hover {
    background-color: #e7c337;
  }
}

.outline {
  padding: calc(var(--space-base) - var(--border-width));
  color: var(--font-color);
  background-color: transparent;
  border: var(--border-width) solid var(--font-color);

  &:hover {
    color: var(--color-black);
    background-color: var(--color-brand-yellow);
    border-color: var(--color-brand-yellow);
  }
}

.invisible {
  color: var(--color-other-blue);
  background-color: transparent;

  &:hover {
    background-color: var(--hover-color);
  }
}

.default {
  color: var(--font-color);
  font-weight: 400;
  background-color: transparent;

  &:hover {
    background-color: var(--hover-color);
  }
}

.block {
  width: 100%;
}

.disabled {
  color: var(--color-gray-5);
  cursor: not-allowed;
  pointer-events: none;
  border-color: var(--color-gray-2);
}

.small {
  padding: var(--space-sm);
}

.medium {
  padding: var(--space-md);
}

.large {
  padding: var(--space-lg);
}

.left {
  .wrapper {
    text-align: left;
  }
}

.center {
  .wrapper {
    text-align: center;
  }
}

.right {
  .wrapper {
    text-align: right;
  }
}

.children {
  width: 100%;
  overflow-wrap: anywhere;
}

.before,
.after {
  display: inline-flex;
}

.before {
  margin-right: var(--space-md);
}

.after {
  margin-left: var(--space-md);
}

.loader {
  position: absolute;
  top: 0;
  right: 15px;
  bottom: 0;
  left: auto;
  height: calc(100% - 20px);
  margin: auto;
  transform-origin: center center;
  animation: rotate 1s linear infinite;
  overflow: hidden;
}

.loaderPath {
  stroke: var(--color-gray-5);
  stroke-dasharray: 93.6375, 124.85;
  stroke-dashoffset: 0;
  stroke-linecap: round;
  stroke-width: 2;
}

@keyframes rotate {
  100% {
    transform: rotate(360deg);
  }
}
