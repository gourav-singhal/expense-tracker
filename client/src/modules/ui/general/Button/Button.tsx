import React from 'react'
import classNames from 'classnames'
import { ButtonVariant, OmittedTypes } from './types'
import s from './Button.module.scss'

interface BaseButtonProps {
  /** React node */
  children?: React.ReactNode
  /**
   * One button variant combinations
   *
   * buttons may be one of a variety of visual variants such as:
   *
   * @default 'primary'
   */
  variant?: ButtonVariant
  /**
   * Renders one of the selected sizes
   *
   * @default base
   * */
  size?: 'small' | 'medium' | 'base' | 'large'
  /**
   * Text alignment inside the button
   *
   * @default center
   * */
  textAlign?: 'left' | 'center' | 'right'
  /**
   * Disables the Button, preventing mouse events,
   * even if the underlying component is an `<a>` element
   */
  disabled?: boolean
  /**
   * Optionally specify a block to render with full width
   *
   * @default false
   * */
  block?: boolean
  /**
   * The loading indicator
   */
  loading?: boolean
  /**
   * The main icon for the button
   *
   * @type ReactNode
   * */
  iconButton?: React.ReactNode
  /**
   * The before icon for the button
   *
   * @type ReactNode
   * */
  before?: React.ReactNode
  /**
   * The after icon for the button
   *
   * @type ReactNode
   * */
  after?: React.ReactNode
  /** Optionally specify a href to render a `<a>` tag styled as a button */
  href?: string
  /**
   *  Set the original html type of button
   *
   *  @default button
   *  */
  type?: 'button' | 'reset' | 'submit'
  /**
   * String passed as HTML class attribute
   */
  className?: string
}

type AnchorButtonProps = BaseButtonProps & Omit<React.AnchorHTMLAttributes<any>, OmittedTypes>

type NativeButtonProps = BaseButtonProps & Omit<React.ButtonHTMLAttributes<any>, OmittedTypes>

type ButtonProps = Partial<AnchorButtonProps & NativeButtonProps>

/**
 * Buttons allow users take actions with a single click.
 * They should be clear indications that they're triggering an action,
 * buttons should be easy to find among other UI.
 */
export const Button: React.FC<ButtonProps> = ({
  children,
  className = '',
  disabled = false,
  variant = 'primary',
  size,
  href,
  block = false,
  iconButton,
  before,
  after,
  type = 'button',
  textAlign = 'center',
  loading = false,
  ...otherProps
}) => {

  const props = {
    className: classNames(s.base, {
        [s.iconButton]: iconButton,
        // variant
        [s.primary]: variant === 'primary',
        [s.outline]: variant === 'outline',
        [s.invisible]: variant === 'invisible',
        [s.default]: variant === 'default',
        // size
        [s.small]: size === 'small',
        [s.medium]: size === 'medium',
        [s.large]: size === 'large',
        // text align
        [s.left]: textAlign === 'left',
        [s.center]: textAlign === 'center',
        [s.right]: textAlign === 'right',
        // state
        [s.block]: block,
        [s.disabled]: disabled || loading
      },
      className
    ),
    disabled,
    type,
    ...otherProps
  }

  const afterContent = loading ? loadingIcon : after

  const contentIconButton = (
    <span className={s.wrapper}>
      {iconButton}
    </span>
  )

  const contentButton = (
    <div className={s.wrapper}>
      {renderBefore(before)}
      {renderChildren(children)}
      {renderAfter(afterContent)}
    </div>
  )

  const innerButton = iconButton ? contentIconButton : contentButton

  if (href) {
    return (
      <a {...props} href={href}>{innerButton}</a>
    )
  }

  return (
    <button {...props}>{innerButton}</button>
  )
}

const loadingIcon = (
  <svg xmlns='http://www.w3.org/2000/svg' className={s.loader} viewBox='25 25 50 50'>
    <circle className={s.loaderPath} cx='50' cy='50' r='20' fill='none' strokeMiterlimit='10' />
  </svg>
)

const renderChildren = (
  children?: React.ReactNode
) =>
  children
    ? <span className={s.children}>{children}</span>
    : null

const renderBefore = (
  before?: React.ReactNode
) =>
  before
    ? <span className={s.before}>{before}</span>
    : null

const renderAfter = (
  after?: React.ReactNode
) =>
  after
    ? <span className={s.after}>{after}</span>
    : null
