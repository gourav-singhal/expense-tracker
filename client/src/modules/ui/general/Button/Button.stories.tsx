import React from 'react'
import { ComponentMeta, ComponentStory } from '@storybook/react'
import { Icon, Space } from 'modules/ui'
import { Button } from './Button'
import 'modules/styles/main.scss'

const iconArrowRight = <Icon icon={'/images/icons/arrow-right.svg'} />
const iconCategories = <Icon icon={'/images/icons/settings/categories.svg'} iconSize={24} />

export default {
  title: 'Design System/General/Button',
  component: Button,
  argTypes: {
    onClick: { action: true },
    type: {
      control: {
        options: ['button', 'reset', 'submit']
      }
    },
    before: {
      control: {
        type: 'text'
      }
    },
    after: {
      control: {
        type: 'text'
      }
    }
  }
} as ComponentMeta<typeof Button>

const Template: ComponentStory<typeof Button> = (args) => <Button {...args} />

export const Default = Template.bind({})

Default.args = {
  children: 'Click me!'
}

export const Primary = Template.bind({})

Primary.args = {
  children: 'Click me!',
  variant: 'primary'
}

export const Outline = Template.bind({})

Outline.args = {
  children: 'Click me!',
  variant: 'outline'
}

export const Invisible = Template.bind({})

Invisible.args = {
  children: 'Click me!',
  variant: 'invisible'
}

export const WithIcon = Template.bind({})

WithIcon.args = {
  children: 'Категории',
  variant: 'default',
  block: false,
  size: 'small',
  textAlign: 'left',
  before: iconCategories,
  after: iconArrowRight
}

const TemplateListLink: ComponentStory<typeof Button> = () => (
  <Space direction='vertical' size={25}>
    <Button variant='primary' href='/'>Click me!</Button>
    <Button variant='outline' href='/'>Click me!</Button>
    <Button variant='invisible' href='/'>Click me!</Button>
  </Space>
)

export const LinkButton = TemplateListLink.bind({})

const TemplateListDisabled: ComponentStory<typeof Button> = () => (
  <Space direction='vertical' size={25}>
    <Button variant='primary' disabled={true}>Don't click me!</Button>
    <Button variant='outline' disabled={true}>Don't click me!</Button>
    <Button variant='invisible' disabled={true}>Don't click me!</Button>
  </Space>
)

export const DisabledButton = TemplateListDisabled.bind({})

const TemplateListLoading: ComponentStory<typeof Button> = () => (
  <Space direction='vertical' size={25}>
    <Button variant='primary' loading={true} size='large'>Don't click me!</Button>
    <Button variant='outline' loading={true} size='base'>Don't click me!</Button>
    <Button variant='invisible' loading={true} size='medium'>Don't click me!</Button>
    <Button variant='default' loading={true} size='small'>Don't click me!</Button>
  </Space>
)

export const LoadingButton = TemplateListLoading.bind({})

export const IconButton = Template.bind({})

IconButton.args = {
  variant: 'invisible',
  iconButton: iconCategories
}
