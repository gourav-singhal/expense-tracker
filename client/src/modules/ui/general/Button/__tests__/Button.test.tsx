import React from 'react'
import renderer from 'react-test-renderer'
import { Button } from '../Button'

describe('Button', () => {
  it('should renders correctly', () => {
    const tree = renderer.create(
      <Button>Click me!</Button>,
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
