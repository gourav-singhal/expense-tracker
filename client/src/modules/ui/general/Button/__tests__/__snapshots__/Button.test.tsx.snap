// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should renders correctly 1`] = `
<button
  className="base primary block"
  disabled={false}
>
  Click me!
</button>
`;
